### PR TITLE
Upsert

### PR DESCRIPTION
**Pull request by johnbos**

### CODE REVIEW

**Review by gerryster — CHANGES_REQUESTED** [review]
Looks good overall, just some minor feedback inline.

File: README.md [hunk 1]
```diff
@@ -24,14 +24,16 @@ Refer to the [Burner](https://github.com/bluemarblepayroll/burner) library for m
...
+* **db_fuel/active_record/update_all** [name, table_name, attributes, debug, register, separator, timestamps, unique_attributes]: This job can take the objects in a register and updates them within a database table.  Attributes defines which object properties to convert to SQL SET clauses while unique_attributes translate to WHERE clauses. One or more records may be updated at a time.  Refer to the class and constructor specification for more detail.
+* **db_fuel/active_record/update** [name, table_name, attributes, debug, register, primary_key, separator, timestamps, unique_attributes]: This job can take the unique objects in a register and updates them within a database table.  Attributes defines which object properties to convert to SQL SET clauses while unique_attributes translate to WHERE clauses to find the records to update. The primary_key is used to update the unique record. Only one record will be updated per statement.  Refer to the class and constructor specification for more detail.
```
[the remark this answers is not in the document]
Comment (johnbos) [reply README.md:30]: Renamed primary_key to primary_keyed_column. Updated references in README.md, also updated note regarding composite primary keys for the update method.

File: lib/db_fuel/modeling/attribute_renderer_set.rb [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    # Also constains methods to transform attribute renderers
+    # and include timestamp attributes if needed.
+    class AttributeRendererSet
+      CREATED_AT = :created_at
+      NOW_TYPE   = 'r/value/now'
+      UPDATED_AT = :updated_at
+
+      attr_reader :attribute_renderers, :resolver
+
+      def initialize(attributes: [], resolver: nil)
```
[the remark this answers is not in the document]
Comment (johnbos) [reply lib/db_fuel/modeling/attribute_renderer_set.rb:22]: Re-ordered parameters and removed default value for resolver.

File: lib/db_fuel/modeling/attribute_renderer_set.rb [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+#
+
+module DbFuel
+  module Modeling
+    # Creates attribute renderers based on attributes passed.
+    # Also constains methods to transform attribute renderers
+    # and include timestamp attributes if needed.
+    class AttributeRendererSet
```
[the remark this answers is not in the document]
Comment (johnbos) [reply lib/db_fuel/modeling/attribute_renderer_set.rb:15]: Added # :nodoc:

File: lib/db_fuel/modeling/attribute_renderer_set.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+      # Adds the attribute for updated_at to the currrent attribute renderers.
+      def timestamp_updated_attribute_renderers
+        timestamp_attributes = [updated_at_timestamp_attribute]
+
+        timestamp_attributes.map do |a|
+          Burner::Modeling::AttributeRenderer.new(a, resolver)
+        end + attribute_renderers
+      end
+
+      def make_attribute_renderers(attributes)
```
[the remark this answers is not in the document]
Comment (johnbos) [reply lib/db_fuel/modeling/attribute_renderer_set.rb:49]: Renamed make_attribute_renderers to make_renderers.

File: lib/db_fuel/modeling/attribute_renderer_set.rb [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        attribute_renderers.each_with_object({}) do |attribute_renderer, memo|
+          value = attribute_renderer.transform(row, time)
+
+          resolver.set(memo, attribute_renderer.key, value)
+        end
+      end
+
+      private
+
+      def created_at_timestamp_attribute
```
[the remark this answers is not in the document]
Comment (johnbos) [reply lib/db_fuel/modeling/attribute_renderer_set.rb:65]: Updated constants for CREATED_AT and UPDATED_AT. Removed timestamp_attribute method.

File: lib/db_fuel/version.rb [hunk 6]
```diff
@@ -8,5 +8,5 @@
 #
 
 module DbFuel
-  VERSION = '1.1.0'
+  VERSION = '1.2.1-alpha.3'
```
[the remark this answers is not in the document]
Comment (johnbos) [reply lib/db_fuel/version.rb:11]: Updated version to 2.0.0.

**Review by gerryster — APPROVED** [review]
LGTM. Note that the created_at_timestamp_attribute and similar updated_at methods can be removed and the constants can be used in their place.

**Pull request merged**